### PR TITLE
Improve offline embedding throughput for pooling models

### PR DESCRIPTION
**Pull request by glaziermag**
TARGETED renderer/tokenization overhead improvement discovered while investigating #41390.

## Problem

For short batched offline embedding workloads, `LLM.embed` can spend avoidable time in renderer-side tokenization before requests reach the engine. In the Nemotron embedding repro from #41390, `BaseRenderer.tokenize_prompts` tokenized each simple text prompt with one `tokenizer.encode(...)` call even when the tokenizer supported list/batch tokenization.

This does not explain the entire vLLM vs. Transformers gap. It is a localized vLLM-side overhead in the steady-state renderer/tokenization path for batched text embedding requests.

## Root Cause

The synchronous renderer path always fell back to:

```python
[self.tokenize_prompt(prompt, params) for prompt in prompts]
```

For plain text prompt lists this misses the tokenizer batch fast path. The Nemotron tokenizer used in the issue does not expose `batch_encode_plus`, but it does support being called with `list[str]`.

## Solution

Add a conservative batch-tokenization fast path for simple text prompt lists:

- use `batch_encode_plus` when available;
- otherwise use the tokenizer callable when it accepts `list[str]`;
- preserve `TokenizeParams` pre-tokenization hooks, encode kwargs, and post-tokenization hooks;
- fall back to the existing per-prompt path for token IDs, prompt embeddings, encoder-decoder inputs, detokenization, unsupported tokenizer APIs, or unexpected return shapes.

The change is general to offline renderer tokenization and does not special-case Nemotron.

## A100 validation update (2026-05-06)

Validated on GCP `a2-highgpu-1g`, 1x NVIDIA A100-SXM4-40GB, driver `580.126.20`, CUDA `13.0`, Ubuntu 24.04, Python `3.12.3`, torch `2.11.0+cu130`, transformers `5.7.0`.

Issue model/workload:

- model: `nvidia/llama-nemotron-embed-1b-v2`
- revision: `cefc2394cc541737b7867df197984cf23f05367f`
- prompts: 5,000 deterministic 32-word query-style prompts with `query: ` prefix
- HF baseline: Transformers `AutoTokenizer` + `AutoModel`, attention-mask mean pooling, `F.normalize`
- vLLM path: offline `LLM.embed`, pooling runner, `FLASH_ATTN`, `max_model_len=8192`

Original issue hardware was 3x A100 80GB PCIe, so this is targeted A100 side-effect validation rather than a full original-environment reproduction.

Correctness on 32 prompts:

- shape: `[32, 2048]`
- HF/vLLM row cosine min/mean/max: `0.999825 / 0.999884 / 0.999911`
- HF norm mean: `1.0`
- vLLM norm mean: `1.0`

Side-effect checks on PR head:

- batch path vs forced old per-prompt fallback: cosine min/mean/max `0.999928 / 0.999998 / 1.000000`, max abs diff `0.001110`;
- direct tokenizer check confirmed batch and per-prompt token IDs are identical for the correctness prompts, so the small embedding delta is runtime numeric variance, not tokenization divergence;
- batch sizes `1/8/32/64/128` returned expected output counts;
- `TokensPrompt` matched text input under the same default special-token semantics;
- mixed `[str, TokensPrompt, str]` inputs worked;
- GPU memory stayed flat after warmup: `4139 MiB` repeated across six samples.

Performance on same A100 and same model revision, count `5000`:

| batch | HF baseline s | vLLM base s | vLLM PR head s | vLLM delta |
| ---: | ---: | ---: | ---: | ---: |
| 1 | 81.150 | 27.597 | 27.601 | 0.0% |
| 8 | 10.814 | 8.910 | 8.781 | 1.5% |
| 32 | 3.106 | 4.775 | 4.258 | 10.8% |
| 64 | 2.776 | 4.139 | 3.629 | 12.3% |
| 128 | 2.620 | 3.913 | 3.375 | 13.7% |

Batch-32 internal profile:

| stage | base ms | PR head ms |
| --- | ---: | ---: |
| `pre_process_offline` | 5.155 | 1.547 |
| `render_and_add_requests` | 2.308 | 3.679 |
| `run_engine` | 23.445 | 23.584 |
| `post_process_offline` | 0.011 | 0.011 |
| total | 32.322 | 30.304 |

The PR improves the local renderer/tokenization overhead and large-batch vLLM wall time, but vLLM remains slower than the HF baseline at larger batches on this single-A100 run. This should be treated as `TARGETED`, not as a full fix for #41390.

## Review follow-up addressed (2026-07-12)

Rebased onto upstream `main` at `27c3e579f0e5f345a86e512e26e1231d3689931f`.

- Checks for a missing tokenizer before capability access.
- Runs pre- and post-tokenization hooks exactly once, including malformed-result and unsupported-API fallback.
- Applies hooks to private prompt copies so fallback does not mutate or double-process the originals.
- Falls back only for an explicit `NotImplementedError` or a structurally invalid batch result; unexpected tokenizer `TypeError` and other implementation failures remain visible.
- Validates the full nested token-ID batch shape and every per-token offset row before accepting the fast path.
- Uses a local fast-tokenizer fixture for hermetic offset tests.

## Tests

- `ruff format --check vllm/renderers/base.py tests/renderers/test_completions.py tests/renderers/test_token_offsets.py`
- `ruff check vllm/renderers/base.py tests/renderers/test_completions.py tests/renderers/test_token_offsets.py`
- `python -m py_compile vllm/renderers/base.py tests/renderers/test_completions.py tests/renderers/test_token_offsets.py`
- `HF_HUB_OFFLINE=1 pytest -q tests/renderers/test_completions.py tests/renderers/test_token_offsets.py --confcutdir=tests/renderers` on CPU -> `42 passed`
- `git diff --check`

The A100 performance/correctness measurements above were made on the earlier fast-path head. This follow-up is safety and compatibility hardening; it does not expand the performance claim.

## AI assistance

This change was developed with AI assistance and reviewed locally before submission.

### CODE REVIEW

**Comment by noooop** [comment]
batch_encode_plus uses multi-threading internally for preprocessing, but it doesn't help much.

https://github.com/vllm-project/vllm/pull/45759

I'm writing a supper-fast offline executor for tiling preprocessing and the engine core.

**Comment by glaziermag** [comment]
Agreed — closing this as superseded. After #45759 and #47699, the offline pooling path renders one request per call through the worker thread pool (`_run_tiling_engine` → `_executor.map(io_processor.render, ...)` → `render_cmpl` with a single-element prompt list), so a batch fast path in `tokenize_prompts` no longer intersects the `LLM.embed` workload this PR targeted. Your TOKENIZERS_PARALLELISM measurement also explains my earlier numbers: the batch-encode gain was internal multithreading, which `renderer_num_workers` plus tiling overlap now provide more generally. Since #47699 fixes #41390, the motivating issue is resolved. If a measurable renderer-side gap remains on current main, I'll open a narrow follow-up instead.
